src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -595,6 +596,7 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
                     link.trim_start_matches(prefix)
                 } else if link.ends_with("!()") {
                     kind = Some(MacroNS);
+                    disambiguator = Some("bang");
```

Comment:
Can we set an enum instead of a string here (yes, it will have multiple variants for each spelling)? Bonus: we can make the `kind` derivation a method on it, so we just call `disambiguator.kind()`, and we can do the same thing with the DefKind, and we can also do the thing with the prefix-matching (something like `fn strip_prefix(&mut String) -> Option<Self>`)

We would have to add a method that converts it back to a useful string though. But we'd be able to nicely move a ton of code out of here.

Reply:
I think we can reuse `DefKind` here. I agree that would be nicer than using strings everywhere, it already has `article()` and `descr()`.

Reply:
@jyn514 I'd rather not use DefKind because we can have multiple disambiguators, like `()` vs `fn` vs `function`.

A cool thing we could do, however, is somehow generate the span for the disambiguator only, and then we don't need to explicitly name it and instead can call it "the function disambiguator here^^". For perf it might be desirable to generate this only when we need to raise an error, i.e. perform a search for it after the fact and adjust the span. Idk.

Reply:
I ended up describing the disambiguator with an article but still naming it . What do you think?

```
error: incompatible link kind for `c`
  --> $DIR/intra-links-disambiguator-mismatch.rs:59:14
   |
LL | /// Link to [c()]
   |              ^^^ help: to link to the constant, use its disambiguator: `const@c`
   |
   = note: this link resolved to a constant, which is not a function
```

Reply:
Oh and if you know how to switch the order of the note and the suggestion please let me know, I'd rather it was
```
error: incompatible link kind for `c`
  --> $DIR/intra-links-disambiguator-mismatch.rs:59:14
   |
LL | /// Link to [c()]
   |              ^^^ this link resolved to a constant, which is not a function
   |
   = help: to link to the constant, use its disambiguator: `const@c`
```